### ISSUE TITLE
Server side is ignoring ?ft= url param

### ISSUE THREAD

**Issue by ericelliott** [paralleldrive/react-feature-toggles#73]
The server side ignores the ?ft= url param, which causes the server to render a 404, but if the user waits a second, the client will take over and render the correct component with the feature enabled.

The server should see the ?ft= url param and send the rendered feature component to the client without a 404 response.

**Comment by nlambert** [comment]
Interesting problem.

**Comment by nlambert** [comment]
We can only do this if `req.url` is supplied. I suggest that react-feature-toggles checks the context to see if `url` has been supplied. It would be up to the project to correctly supply url to the context.

`withFeatures()` is setting `windowLocationSearch` in a closure. Instead `withFeaturesHOC` could grab `url` from context in its constructor. If its not there, then it could try `window.location.search`.

I don't know this project very well. Maybe there is a better way more in line with the current direction. @kennylavender any thoughts?

**Comment by kennylavender** [comment]
Can you give me an example of the problem? React feature toggles checks req.url in [create-route-middleware](https://github.com/paralleldrive/react-feature-toggles/blob/master/src/create-route-middleware/index.js) so that you can render a specific component for a 404 or 200 on the server side that matches the client side.

The [nextjs example](https://github.com/paralleldrive/react-feature-toggles/blob/master/examples/nextjs/server.js)

**Comment by nlambert** [comment]
I'll have a look. The project might not be using that middleware. Thanks.

**Comment by nlambert** [comment]
@kennylavender does `createRoute()` update context for use in `configureFeature()`? do you have an example of `createRoute()` usage?

**Comment by kennylavender** [comment]
`withFeatures` sets the context for `configureFeature`.

`createRoute` the returned function from `createRouteMiddelware` does not use or set context. Its a server side only function for creating an express handler that checks if a feature is enabled or disabled, if enabled then it sets `req.status` to 200, else it sets `req.status` to 400.

The nextjs example above is the example for `createRouteMiddleware`

**Comment by nlambert** [comment]
@kennylavender you supplied an example. missed that.

I'll start with that. Thanks.

**Comment by kennylavender** [comment]
Cool sounds good, I updated my comment above a bit to clarify and fix some stuff. Make sure to read the update version.

The nextjs example is letting nextjs decide which component to render because nextjs already has the logic built in to render a 404 component if `res.status` is 404.

**Comment by nlambert** [comment]
Implemented `createRouteMiddleware()`, but it's not solving the issue. It did make the code nice and clean though ;-)

The description above is a little misleading. There isn't a 404, but I wouldn't blame anyone experiencing the isue for thinking there was.

What is happening:
```
const error404 = () => <Error statusCode={404} />;
const featureOr404 = configureFeature(error404);
```

`configureFeature()` is just rending the `FallbackComponent` on the server. It works properly is the feature is enabled in the features file. I can only assume `withFeatures()` is not correctly updating the context when `ft=` is used.

**Comment by ericelliott** [comment]
@kennylavender Is there a working example in the tests we can refer to which renders the correct response given the url params?

**Comment by kennylavender** [comment]
@nlambert Yeah, the nextjs Error component is what I meant.

@ericelliott here is the [integration test for create-route-middleware](https://github.com/paralleldrive/react-feature-toggles/blob/master/src/test/integration/create-route-middleware.js)

@nlambert not quite sure I understand the issue. Let me re-read what you said a few times, it might sink in. I did have the server side rendering and client side working together correctly in another project. But your configuration may be different.

**Comment by kennylavender** [comment]
@ericelliott here is a [function test https://github.com/paralleldrive/echo-and-the-geckos/blob/master/source/server/routes/test/profile.js and pr https://github.com/paralleldrive/echo-and-the-geckos/pull/254

**Comment by kennylavender** [comment]
@nlambert can you give me a link to the server code.

**Comment by nlambert** [comment]
@kennylavender 
1. hit the server with ?ft=feature-name
2.`createRouteMiddleware()` correctly determines feature is enabled via params
3. server rendering handed to nextjs
4. page is wrapped with `withFeatures()`. `initialFeatures` passed in
5. configureFeature(error404)('feature-name')(FeatureComponent) renders the fallback, not the FeatureComponent

Hope this helps

**Comment by kennylavender** [comment]
Yeah that does, ok thats what I thought you were getting at.

**Comment by nlambert** [comment]
Link to server code https://goo.gl/mVekbK

**Comment by kennylavender** [comment]
Ah ok got it, and replicated it on the other project.

**Comment by kennylavender** [comment]
@nlambert Your initial comments were correct

**Comment by kennylavender** [comment]
Il have to think about how to provide the correct features or url to with-features while its rendering on the server.

**Comment by kennylavender** [comment]
Using context seems like a good way, like you said @nlambert. Il let you know if I can think of anything else.

**Comment by nlambert** [comment]
Thinking about context collisions now. `url` seems like a collision waiting to happen and wrapping everything is a breaking change.

**Comment by ericelliott** [comment]
I'm confused. Why would we need another channel to pass the query?

`query` is passed as a param to `getInitialProps()`, which should then be passed through the HOC props.

```js
const Page = ({ query }) => (
  <div>
    { `Query is: ${query}` }
  </div>
)

Page.getInitialProps = ({ query }) => {
  return { query: JSON.stringify(query) }
}

export default Page
```

So it should be available to us as a prop, assuming that:

1. The top level component implements `getInitialProps` and passes through the `query` prop. (We should document this requirement).
2. We use that `query` prop to calculate the available features in [with-features](https://github.com/paralleldrive/react-feature-toggles/blob/d49c49c9a1ffc86356c5658f8b98d7a1f8544b6b/src/with-features.js#L28). We don't seem to be doing that, yet.

**Comment by nlambert** [comment]
We would be using that `query` provided to `getInitialProps()` and pass it to context.

The issue with `getInitialProps()` is that it is only called by nextjs on the outermost component. If `withFeatures()` is called a few levels down we need a way to reliably get the data there. That way `withFeatures()` doesn't need to be the outermost component (although it could): 

```
export default compose(
  withInitialProps(callbackFunc),
  withRedux,
  withEnv,
  withFeatures,
  withFirebase,
  withMaterialUI,
  withLayout
);
``` 

In the example `withInitialProps()` could be an hoc that implements `getInitialProps`. Its the only one that can since its the outermost hoc. The `callbackFunc()` will receive what nextjs gives it and can supply `context` with data like `query`.

`withFeatures()` will call `getEnabledFeatures()` with the _search_ key of the query that it gets from `context`.

**Comment by kennylavender** [comment]
I was thinking along the same lines as @nlambert, I figured setting context might be better than passing the props through the HOC' chain, but props would be easier to implement. So I am up for either way.

**Comment by ericelliott** [comment]
1. All HOCs should pass through props.
2. Any top-level Page HOC should implement getInitialProps and pass query through. This is an established practice in the Next.js docs, which I quoted in the example code above.
3. We are already passing features as both props and context. This component doesn't need to pass query anywhere... just take it as a prop and use it in its features calculation.

It's not our job here to supply query. That's out of scope.

We can specify that the query prop supplied by getInitialProps is a dependency, and make that clear in the docs.

**Comment by nlambert** [comment]
1. Yes
2. Yes
3. Yes (ok just noticed that features is _also_ passed as props)

I think I'm seeing the global picture here:
- `configureFeature()` is the only other React component other than `withFeatures()`
- `configureFeature()` can be used anywhere in the project and depends on features being supplied in `context`
- `withFeatures()` is a top level hoc and can grab query from `props`
- `withFeatures()` should grab `query` from props and pass `query.search` to `getEnabledFeatures()`. The fallback is `window.location.search` or empty string
- Update documentation: condition to using the package is that you pass `query` as props

**Comment by kennylavender** [comment]
@nlambert sounds good to me! Let me know if I can help at all.

**Comment by ericelliott** [comment]
![make_it_so](https://user-images.githubusercontent.com/364727/34082634-c8c86d00-e316-11e7-86b1-5be35fc9f97c.jpg)

**Comment by kennylavender** [comment]
Should be complete with last PR merge

**Comment by ericelliott** [comment]
Published as v1.3.0
